src/utils/display-observer.js

```diff
@@ -0,0 +1,226 @@
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
+      }
+    );
+
+    /** @private @const {!IntersectionObserver} */
+    this.viewportObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, 1),
+      {threshold: 0.001}
+    );
+
+    /** @private {boolean} */
+    this.isDocVisible_ = ampdoc.isVisible();
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() => {
+      if (ampdoc.isVisible() !== this.isDocVisible_) {
+        this.isDocVisible_ = ampdoc.isVisible();
+        this.docVisibilityChanged_();
+      }
+    });
+
+    /** @private @const {!Map<!Element, !Array<function(boolean)>>} */
+    this.targetObserverMultimap_ = new Map();
+
+    /** @private @const {!Map<!Element, !Array<boolean>>} */
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverMultimap_.set(target, callbacks);
+      this.docObserver_.observe(target);
+      this.viewportObserver_.observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      const displays = this.targetDisplayMap_.get(target);
+      if (displays) {
+        setTimeout(() =>
+          callCallbackNoInline(
+            callback,
+            getDisplay(displays, this.isDocVisible_)
+          )
+        );
+      }
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function()} callback
+   */
+  unobserve(target, callback) {
+    const callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.targetObserverMultimap_.delete(target);
+      this.targetDisplayMap_.delete(target);
+      this.docObserver_.unobserve(target);
+      this.viewportObserver_.unobserve(target);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targetObserverMultimap_.forEach((callbacks, target) => {
+      const displays = this.targetDisplayMap_.get(target);
+      const oldDisplay = getDisplay(displays, !this.isDocVisible_);
+      const newDisplay = getDisplay(displays, this.isDocVisible_);
+      if (newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {number} index
+   * @private
+   */
+  observed_(entries, index) {
```

Comment:
Can we make `index` an enum or at least named const?

Reply:
Good idea. Done.